core/src/main/scala/kafka/server/ZkIsrManager.scala

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+package kafka.server
+
+import kafka.utils.{Logging, ReplicationUtils, Scheduler}
+import kafka.zk.KafkaZkClient
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.utils.Time
+
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicLong
+import scala.collection.mutable
+
+/**
+ * @param checkIntervalMs How often to check for ISR
+ * @param maxDelayMs  Maximum time that an ISR change may be delayed before sending the notification
+ * @param lingerMs  Maximum time to await additional changes before sending the notification
+ */
+case class IsrChangePropagationConfig(checkIntervalMs: Long, maxDelayMs: Long, lingerMs: Long)
+
+object ZkIsrManager {
+  // This field is mutable to allow overriding change notification behavior in test cases
+  @volatile var DefaultIsrPropagationConfig: IsrChangePropagationConfig = IsrChangePropagationConfig(
+    checkIntervalMs = 2500,
+    lingerMs = 5000,
+    maxDelayMs = 60000,
+  )
+}
+
+class ZkIsrManager(scheduler: Scheduler, time: Time, zkClient: KafkaZkClient) extends AlterIsrManager with Logging {
+
+  private val isrChangeNotificationConfig = ZkIsrManager.DefaultIsrPropagationConfig
+  // Visible for testing
+  private[server] val isrChangeSet: mutable.Set[TopicPartition] = new mutable.HashSet[TopicPartition]()
+  private val lastIsrChangeMs = new AtomicLong(time.milliseconds())
+  private val lastIsrPropagationMs = new AtomicLong(time.milliseconds())
+
+  override def start(): Unit = {
+    scheduler.schedule("isr-change-propagation", maybePropagateIsrChanges _,
+      period = isrChangeNotificationConfig.checkIntervalMs, unit = TimeUnit.MILLISECONDS)
+  }
+
+  override def clearPending(topicPartition: TopicPartition): Unit = {
+    // Since we always immediately process ZK updates and never actually enqueue anything, there is nothing to
+    // clear here so this is a no-op. Even if there are changes that have not been propagated, the write to ZK
+    // has already happened, so we may as well send the notification to the controller.
+  }
+
+  override def enqueue(alterIsrItem: AlterIsrItem): Boolean = {
+    debug(s"Writing new ISR " + alterIsrItem.leaderAndIsr.isr + " to ZooKeeper with version " +
+      alterIsrItem.leaderAndIsr.zkVersion + " for partition " + alterIsrItem.topicPartition)
+
+    val (updateSucceeded, newVersion) = ReplicationUtils.updateLeaderAndIsr(zkClient, alterIsrItem.topicPartition,
+      alterIsrItem.leaderAndIsr, alterIsrItem.controllerEpoch)
+
+    if (updateSucceeded) {
+      // Track which partitions need to be propagated to the controller
+      isrChangeSet synchronized {
+        isrChangeSet += alterIsrItem.topicPartition
+        lastIsrChangeMs.set(time.milliseconds())
+      }
+
+      // We actually need to apply the callback in another thread since Partition#sendAlterIsrRequest will write
```

Comment:
Hmm.. It feels a tad brittle to rely on locking assumptions like this. Perhaps we could let `enqueue` return a different response in the case that the change was applied immediately? For example, we could return `Option[Int]` to indicate that new version if the change was applied. Note that we can probably discard the boolean return type and let `enqueue` raise an illegal state exception directly since that is what `Partition` is already doing.

Reply:
Ok, so in this case, we would need to call the callback from `sendAlterIsrRequest` for the ZK case. I think that would probably be ok. I agree that relying on the separate thread and the lock is a little wonky.

Something like this?
```scala
try {
  alterIsrManager.enqueue(alterIsrItem) match {
    case Some(newVersion) => 
      isrState = proposedIsrState
      handleAlterIsrResponse(proposedState.withZkVersion(newVersion))
    case None =>
      isrState = proposedIsrState
} catch {
   e: Throwable => 
    isrChangeListener.markFailed()
    throw e
}
```

Reply:
One issue is the invalid version case for ZK -- if we return None there then partition will think it's a call to DefaultAlterIsrManager... I guess we could throw from ZkIsrManager if the update fails. WDYT?

Reply:
I went with a slightly different approach. See the latest commits